runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/config/CompressionClientConfig.kt

```diff
@@ -0,0 +1,92 @@
+import aws.smithy.kotlin.runtime.compression.Gzip
+
+@DslMarker
+public annotation class CompressionClientConfigDsl
+
+public interface CompressionClientConfig {
+    public val requestCompression: RequestCompressionConfig
+
+    @CompressionClientConfigDsl
+    public interface Builder {
+        public fun requestCompression(block: RequestCompressionConfig.Builder.() -> Unit) {
+            requestCompression.apply(block)
+        }
+
+        public val requestCompression: RequestCompressionConfig.Builder
+    }
+}
+
+/**
+ * The configuration properties for request compression.
+ */
+public class RequestCompressionConfig(builder: Builder) {
+    public companion object {
+        public inline operator fun invoke(block: Builder.() -> Unit): RequestCompressionConfig =
+            RequestCompressionConfig(Builder().apply(block))
+    }
+
+    @InternalApi
+    public fun toBuilderApplicator(): Builder = Builder().apply {
+        compressionAlgorithms = this@RequestCompressionConfig.compressionAlgorithms.toMutableList()
+        disableRequestCompression = this@RequestCompressionConfig.disableRequestCompression
+        requestMinCompressionSizeBytes = this@RequestCompressionConfig.requestMinCompressionSizeBytes
+    }
+
+    /**
+     * The list of compression algorithms supported by the client.
+     * More compression algorithms can be added and may override an existing implementation.
+     * Use the `CompressionAlgorithm` interface to create one.
+     */
+    public val compressionAlgorithms: List<CompressionAlgorithm>
+
+    /**
+     * Flag used to determine when a request should be compressed or not.
+     * False by default.
+     */
+    public val disableRequestCompression: Boolean
+
+    /**
+     * The threshold in bytes used to determine if a request should be compressed or not.
+     * MUST be in the range 0-10,485,760 (10 MB). Defaults to 10,240 (10 KB).
+     */
+    public val requestMinCompressionSizeBytes: Long
+
+    /**
+     * A builder for [CompressionClientConfig]
+     */
+    @CompressionClientConfigDsl
+    public class Builder {
+        /**
+         * The list of compression algorithms supported by the client.
+         * More compression algorithms can be added and may override an existing implementation.
+         * Use the `CompressionAlgorithm` interface to create one.
+         */
+        public var compressionAlgorithms: MutableList<CompressionAlgorithm> = mutableListOf(Gzip())
+
+        /**
+         * Flag used to determine when a request should be compressed or not.
+         * False by default.
+         */
+        public var disableRequestCompression: Boolean? = null
+
+        /**
+         * The threshold in bytes used to determine if a request should be compressed or not.
+         * MUST be in the range 0-10,485,760 (10 MB). Defaults to 10,240 (10 KB).
+         */
+        public var requestMinCompressionSizeBytes: Long? = null
+    }
+
+    init {
```

Comment:
**style**: Move this to top of the class before the companion object.

Reply:
I get an error saying `Variable cannot be initialized before declaration`. The highest it can be is under the variable declarations, is there a way around this? If I do just place it under the variable declarations do I refactor the `toBuilderApplicator` and companion object to be somewhere else?

Reply:
You don't even need the init block you could just assign them directly

```kotlin

    public val compressionAlgorithms: List<CompressionAlgorithm> = builder.compressionAlgorithms

    ...

    public val disableRequestCompression: Boolean = builder.disableRequestCompression ?: false

    ...
```